### PR TITLE
WIP: Sketch of possible design for 3.x

### PR DESCRIPTION
**Pull request by jdegoes**
Together with @adamfraser, we created a sketch of a possible design for Monocle 3.x.

The design:

 - **Reflects the optic hierarchy using subtyping**. Now an IDE can tell users that a lens is an optional, for example, without users having to remember or use conversion functions.
 - **Fully embraces declaration-site variance**. To use declaration-site variance, the fully polymorphic optic types are required; this, in general, will _improve_ type inference, despite the additional number of parameters.
 - **Simplifies implementations**. By making many methods `final`, duplication across optic types can be eliminated with default implementations that are comparable (allocation and dispatch-wise) to the specialized alternatives previously present.
 - **Removes all dependencies**. This pull request uses a simplified encoding for `Traversal`, which should encapsulate 99% of all use cases. It's worth debating whether or not to introduce a `Combine` interface (without implicits) that features `zipWith` and `make` methods to zip two structures together and make one from a value (which is wholly equivalent in terms of expressive power to the current encoding and can enable separate `interop` libraries to create these optics for any traversable).

We are not done but this is good enough for some initial feedback. Remaining work:

 - Perfect type inference so users can compose all optics using the same name, e.g.:
    ```scala
    _Org >>> _Site >>> _Manager >>> _Employees >>> _Salary
    ```
    Not having to choose a method name based on the target composition type will dramatically improve usability of the API and appearance of usage, and we have done a (somewhat unsatisfying) POC to show it's possible, but we look for better ways of doing it.
 - Currently using `Lens.first` and so forth without overloads results in type inference issues. Adam and I discussed how to solve this problem using a technique introduced in ZIO Test. It's quite possible this would result in flawless type inference in all cases at some cost for library maintainers. We have to POC it to see if it works and if the overhead is worth it.

Finally, there are a huge number of combinators that can be defined on these optics, which are not currently defined in Monocle 2.x, which can make the library much more powerful and user-friendly out of the box. However, before defining them or at least sketching them out, I think it's worth solidifying the core design.

### CODE REVIEW

**Comment by julien-truffaut** [comment]
Thank you, John and Adam, for the pull request. I will need some times to go through it in details. Please find below my initial comments:

It is great you are trying to tackle the issue of non-overloaded compose methods (i.e. removing `composeLens`, `composePrism`). We tried many approaches over the year unsuccessfully. Your attempt looks quite similar to a spike I did with Edward Kmett a couple of years ago. The bottom line is that it is a bug/missing feature in scala 2 (https://github.com/julien-truffaut/Monocle/issues/417), and I verified it is fixed in dotty (https://github.com/julien-truffaut/Monocly).

You make a good case to keep the polymorphic optics. I don't think polymorphic optics are practical in Scala, but at least they would allow variance for Getter and Fold. I am not sure variance will improve type inference in the monomorphic case since type parameters are invariant in that scenario.

I also had a version with sealed hierarchy, but I ended up not going for it. In my opinion, an optic is an interface, not an enumeration. If we sealed the hierarchy, it would allow people to pattern match on optics which would break encapsulation. For example, if a function takes a Getter as an input, one could check if the Getter is a Lens or Iso and in that case, they could use set/reverseGet to modify. I am curious to know your arguments for sealing the hierarchy. 

Great implementation for Fold. I need to take more time to review Traversal. Could you elaborate on the limitations of your encoding (you seem to imply it has some)? For example, can you use a Traversal with Validated (to accumulate errors) or with IO?

I like `>>>` as an operator alias for `compose`. It looks like `>>>` is an alias for `andThen` in `Category`  maybe we should also rename optics `compose` function to `andThen` to be consistent.

Cheers

File: kernel/shared/src/main/scala/monocle/Iso.scala [hunk 1]
```diff
@@ -2,37 +2,13 @@ package monocle
...
-  def apply[A, B](_get: A => B)(_reverseGet: B => A): Iso[A, B] =
-    new Iso[A, B] {
-      def get(from: A): B      = _get(from)
-      def reverseGet(to: B): A = _reverseGet(to)
-    }
+  def apply[S, A](_get: S => A)(_reverseGet: A => S): Iso[S, A] =
+    PIso[S, S, A, A](_get)(_reverseGet)
 
-  def reverse[A, B](implicit ev: Reverse.Aux[A, B]): Iso[A, B] =
+  def reverse[S, A](implicit ev: Reverse.Aux[S, A]): Iso[S, A] =
```
Comment (cquiroz) [inline kernel/shared/src/main/scala/monocle/Iso.scala]: I think we standardized on using `A, B, ..` for type params rather than `S, A`

Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/Iso.scala]: We might need to revisit that if we do go for polymorphic optics. We could have polymorphic optics with A1, A2, B1, B2, e.g. `PLens[A1, A2, B1, B2]` and `Lens[A, B]`

File: kernel/shared/src/main/scala/monocle/optics.scala [hunk 2]
```diff
@@ -0,0 +1,197 @@
+package monocle
+
+sealed trait Fold[-S, +A] extends Serializable {
+  def fold[Z](zero: Z, combine: (Z, Z) => Z)(f: A => Z)(s: S): Z
+}
+
+sealed trait PFold[+LowerS, -UpperS, +LowerA, -UpperA] extends Serializable {
```
Comment (julien-truffaut) [inline kernel/shared/src/main/scala/monocle/optics.scala]: Do we need PFold?

File: kernel/shared/src/main/scala/monocle/optics.scala [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  def apply[S, T, A, B](modify0: (S, A => B) => T)(set0: (S, B) => T): PSetter[S, T, A, B] =
+    new PSetter[S, T, A, B] {
+      def modify(f: A => B): S => T = modify0(_, f)
+      def set(b: B): S => T = set0(_, b)
+
+    }
+
+}
+
+sealed trait PTraversal[S, T, A, B] extends PSetter[S, T, A, B] { self =>
```
Comment (julien-truffaut) [inline kernel/shared/src/main/scala/monocle/optics.scala]: PTraversal should extend Fold[S, A] as well

File: kernel/shared/src/main/scala/monocle/optics.scala [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                (ez, c) =>
+                  ez.map(f(_, c)) match {
+                    case Left(z) => Left(Left(z))
+                    case Right(a) => a match {
+                      case Left(z) => Left(Left(z))
+                      case Right((z, d)) => Right((Right(z), d))
+                    }
+                  }
+              )(a) match {
+                case (Left(z), b) => ???
```
Comment (julien-truffaut) [inline kernel/shared/src/main/scala/monocle/optics.scala]: ???

**Comment by julien-truffaut** [comment]
@adamgfraser @jdegoes when you have time, I would love to hear your comments.

File: kernel/shared/src/main/scala/monocle/optics.scala [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    }
+
+}
+
+sealed trait PTraversal[S, T, A, B] extends PSetter[S, T, A, B] { self =>
+  def traverse[Z](z: Z, f: (Z, A) => Either[Z, (Z, B)]): S => (Z, T)
+
+  final override def fold[Z](zero: Z, combine: (Z, Z) => Z)(f: A => Z)(s: S): Z =
+    traverse[Z](zero, (z, a) => Left(combine(z, f(a))))(s)._1
```
Comment (julien-truffaut) [inline kernel/shared/src/main/scala/monocle/optics.scala]: I started to play a bit with this encoding. I am not entirely sure what does it mean to return a Left or a Right in traverse. At first, I thought Left was a way to short cut the traversal but it doesn't look to be the case.

@jdegoes @adamgfraser do you have an example of a `Traversal` for all elements in a List.

**Comment by jdegoes** [comment]
@julien-truffaut I've been extraordinarily busy and not had a chance to return to this, but will as soon as I can. Also @adamgfraser and I are working on a radically different encoding that uses one type for everything but we still have work to do.

**Comment by julien-truffaut** [comment]
Closing this as we are making experiment in Monocly reposoitory. But lots of good ideas from @jdegoes and @adamgfraser, thanks.

**Pull request closed without merging**